twitter-as.js

```diff
@@ -0,0 +1,226 @@
+        url: uri,
+        remote: new RemoteUser(twitter_user.id_str)
+      };
+    });
+  }).catch((error) => {
+      console.error("Couldn't find the user.");
+      console.error("Reason:", error);
+  });
+}
+
+function roomQuery(alias, aliasLocalpart) {
+    console.log(aliasLocalpart);
+    return Twitter.get_user(aliasLocalpart.substr("@twitter_".length)).then((tuser) => {
+        console.log(tuser);
+        if (tuser != null) {
+            if (!tuser.protected) {
+                return constructTimelineRoom(tuser, aliasLocalpart);
+            }
+        }
+    });
+}
+
+function eventQuery(request, context) {
+    var event = request.getData();
+    console.log(event.type);
+    if (event.type == "m.room.member") {
+        if (event.membership == "invite" && event.state_key.startsWith("@twitter_")) { //We should prolly use a regex
+            var intent = bridge.getIntent(event.state_key);
+            intent.join(event.room_id);
+
+            //Set the avatar based on the 'owners' avatar.
+            intent.getClient().getProfileInfo(event.state_key, 'avatar_url').then((url) => {
+                console.log("Set Room Avatar:", url);
+                intent.sendStateEvent(event.room_id, "m.room.avatar", "", {
+                    "url": url.avatar_url
+                });
+            });
+
+            if (context.rooms.remote != null) {
+                Twitter.add_timeline(event.state_key, context.rooms.matrix, context.rooms.remote);
+            } else {
+                console.log("Couldn't find the remote room for this timeline.");
+            }
+        }
+    }
+
+    //console.log("Request:",request);
+    //console.log("Context:",context);
+    return; // we will handle incoming matrix requests later
+}
+
+/*
+  This will create a stream room for one users timeline.
+  Users who are not authenticated via OAuth will recieve the default power of 0
+  Users who are authenticated via Oauth will recieve a power level of 10
+  The owner of this stream will recieve a 75
+  The bot will have 100
+*/
+function constructTimelineRoom(user, aliasLocalpart) {
+    var botID = bridge.getBot().getUserId();
+
+    var roomOwner = "@twitter_" + user.id_str + ":" + bridge.opts.domain;
+    var users = {};
+    users[botID] = 100;
+    users[roomOwner] = 75;
+    var powers = roomPowers(users);
+    var remote = new RemoteRoom("timeline_" + user.id_str);
+    remote.set("twitter_type", "timeline");
+    remote.set("twitter_user", roomOwner);
+    opts = {
+        visibility: "public",
+        room_alias_name: aliasLocalpart,
+        name: "[Twitter] " + user.name,
+        topic: user.description,
+        invite: [roomOwner],
+        initial_state: [
+            powers, {
+                "type": "m.room.join_rules",
+                "content": {
+                    "join_rule": "public"
+                },
+                "state_key": ""
+            }, {
+                "type": "org.twitter.data",
```

Comment:
Are you Twitter? No, you're not. So please don't use `org.twitter`'s namespace. `org.matrix.twitter.data` would be fine, as would `uk.halfshot.twitter.data`.